src/Microsoft.AspNetCore.Mvc.Razor.Extensions/RazorPageDocumentClassifierPass.cs

```diff
@@ -10,6 +10,21 @@ namespace Microsoft.AspNetCore.Mvc.Razor.Extensions
     public class RazorPageDocumentClassifierPass : DocumentClassifierPassBase
     {
         public static readonly string RazorPageDocumentKind = "mvc.1.0.razor-page";
+        private static readonly RazorEngine LeadingDirectiveParsingEngine = RazorEngine.Create(builder =>
+        {
+            for (var i = builder.Phases.Count - 1; i >= 0; i--)
+            {
+                var phase = builder.Phases[i];
+                builder.Phases.RemoveAt(i);
+                if (phase is IRazorDocumentClassifierPhase)
+                {
+                    break;
+                }
+            }
+
+            RazorExtensions.Register(builder);
+            builder.Features.Add(new LeadingDirectiveParserOptionsFeature());
```

Comment:
Let me get out in front of this and say that I really don't think us 'creating' an engine is something we want to live with. Adding something like this is going to just cause lots of complexity for us down the line.

Could we just examine the IR tree and say "good enough". The user feedback we've gotten about this problem is aimed at  catching obvious mistakes.

Reply:
Or could we call the static methods on the parser (they exist for a reason).

Reply:
> Could we just examine the IR tree and say "good enough". The user feedback we've gotten about this problem is aimed at catching obvious mistakes.

IR doesn't represent Razor comments.

> Or could we call the static methods on the parser (they exist for a reason).

They're internal, we'd have to IVT for that.

> Let me get out in front of this and say that I really don't think us 'creating' an engine is something we want to live with. Adding something like this is going to just cause lots of complexity for us down the line.

Ehh, in this case i'd argue it reduces the complexity because this is what MVC does. Basically we get identical behavior to MVC for free without having to mock the same behavior.

Reply:
> IR doesn't represent Razor comments.

We don't care about comments in this case, do we?

Reply:
> We don't care about comments in this case, do we?

The single line parser does (it ignores them)

Reply:
> We don't care about comments in this case, do we?

We want to allow comments to be present before the `@page`.

Reply:
Sure. This is the bit I don't follow:

> IR doesn't represent Razor comments.

The presence or absence of comments should be immaterial here. We just need to make sure, no other sematically meaningful entry exists before `@page`. So wouldn't the IR tree suffice for this purpose?

Reply:
> The presence or absence of comments should be immaterial here.

The problem is with Razor comments are represented as CSharpCode just like any other content. How will you differentiate a semantically meaningful entry from a Razor comment if they are both represented by the same node? Looking into it's content will get ugly and complicated.

Reply:
> The problem is with Razor comments are represented as CSharpCode just like any other content.

Gotcha. This I did not know. I assumed it was missing those nodes in entirety.